src/Compiler.Dynamic/SystemWebCompilation.cs

```diff
@@ -169,13 +167,23 @@ private async Task<T> InternalCompilePageAsync(string path, CancellationToken to
 
             var compiled = CreateCompiledPage(compilation, path, typeName, trees, references!, embedded, assemblies!, token);
 
+            if (path.EndsWith(".ascx"))
+            {
+                var userControlType = compiled.Type;
+                if (userControlType is null && compiled is PersistedCompiledPage persistedCompiledPage)
+                {
+                    var loadedAssembly = Assembly.LoadFile(persistedCompiledPage.FullPath);
+                    userControlType = loadedAssembly.GetType(typeName);
+
+                }
```

Comment:
if there is a better way to tap into to assembly/type during static compilation, suggest...

Reply:
definitely don't like this. Why do you need this?

Reply:
Took a look at where it's used - I'd prefer a new internal type such as

```
internal interface IUserControlProvider
{
Type? GetUserControl(string virtualPath);
}
```

Then the compiler can implement it and track the compilations